### DISCUSSION TITLE
Server component ( parent ) is controlling Client component ( child )

### DISCUSSION THREAD

**Discussion by AndonMitev** [vercel/next.js#60463] [Help]
### Summary

So i do have 2 components, 1 Parent - server component, table row && 1 Child button component.

Child component is not receiving anyprops however it has console.log and sometimes i can see this log happening on the server console. Here is the child Component:

```
'use client';

import { EmitterEvents, ExchangeTrendsActions, ModalTypes } from '@/constants';
import { useIsChainSupported } from '@/hooks';
import { ExchangeButtonProps } from '@/types';
import { eventEmitter } from '@/utils';
import Button from './Button';

export default function BuyButton({ trend, ...rest }: ExchangeButtonProps) {
  const isChainSupported = useIsChainSupported();

  const handleClick = () => {
    eventEmitter.emit(EmitterEvents.ToggleModal, {
      modalType: ModalTypes.ExchangeTrendsModal,
      data: {
        ...trend,
        action: ExchangeTrendsActions.Buy,
      },
    });
  };

  console.log('what', isChainSupported);

  return (
    <Button disabled={!isChainSupported} onClick={handleClick} {...rest}>
      Buy
    </Button>
  );
}

```

Sometimes `what` is being executed on the server i also can see logs in the console, however `disabled`  is taken state from the server instead from the client, component is marked as a client, why is this happening, any ideas?

Log in server console:
<img width="698" alt="image" src="https://github.com/vercel/next.js/assets/32274987/a8b574f2-6221-4aa4-a3c4-112b4778ed01">

Also log from browser console:
<img width="328" alt="image" src="https://github.com/vercel/next.js/assets/32274987/4e1a2a5b-7194-4a2d-846f-b7b61108389f">

However button are disabled, like server is with priority, any idea how to prevent server behaviour?

**Comment by icyJoseph** [comment]
Hi,

Unfortunately, many tech influencers, hastily written articles, and the choice of `use client` as the directive name, have caused people to think that, Client Component are only rendered in the Browser. 

That's not true. Client Components are SSR'd:

- https://github.com/reactwg/server-components/discussions/4

It is often more useful to think of the, `use client` directive, as hint to the bundler. And we hint that, this JavaScript file, module, etc, should be packed and sent to the client. Of course, this is also a hastily written simplification, and it is only part of what's really happening.

This is what gives us the ability to do things in the browser, AKA, do interactivity handling. 

This is all documented, and said several times in the Next.js and React docs though.

**Reply by AndonMitev** [reply]
Yeah @icyJoseph I've completely agree with your point, i know each component is pre-rendered on the server doesn't matter if client or server. Here is something very new and interesting by Dan: https://overreacted.io/the-two-reacts . 

Ok i understood that button is pre-rendered on the server and after that send to the client for hydratation etc, but can't understand why is not respecting state changes that are happening on the client, i mean i can see in browser console how boolean is updated to `true` but component it self is not triggering update and button stay disabled:

<img width="843" alt="image" src="https://github.com/vercel/next.js/assets/32274987/7a46f6a6-994f-4f00-9247-1b52ebef3d26">

Any advice what can be useful in such situation?

**Reply by AndonMitev** [reply]
Yeah i can confirm with `ssr: false` works but adds nasty delay to the ui until component is shown ❌ 
```
export default dynamic(() => Promise.resolve(BuyButton), {
  ssr: false,
  loading: () => <></>,
});
```

If there is better approach i'd love to hear it

**Reply by icyJoseph** [reply]
Well, what's in the `rest` object? Are you maybe forwarding a disable coming from the upper parent? 

Show us what the React Dev Tools show for Button props.

This is most likely an error on your side.

**Reply by AndonMitev** [reply]
Sure this are the props

<img width="559" alt="image" src="https://github.com/vercel/next.js/assets/32274987/ad499504-a42e-480c-97ef-ab41f6d8604e">


here is the interface about em:

```
export type ExchangeButtonProps = {
  variant?: ButtonVariants;
  size?: ButtonSizes;
  trend: {
    trendName: string;
    trendHash: string;
    trendPrice: bigint;
    shareSupply: number;
  };
};
```

There is no disable/enable state passing from parent, this state is managed entirely from the Button it self

**Reply by icyJoseph** [reply]
What does React Dev Tools show for Button? Could you also expand the `IsChainSupported` hook?

**Reply by AndonMitev** [reply]
Yeah this is the state of underlying button

BuyButton:
<img width="702" alt="image" src="https://github.com/vercel/next.js/assets/32274987/c668a2d9-2df4-4586-a643-9854e9ed047f">


SellButton:
<img width="666" alt="image" src="https://github.com/vercel/next.js/assets/32274987/f81ba96e-3f11-4b3e-adb2-59663288014e">

Keep in mind I've have reverted:
```
export default dynamic(() => Promise.resolve(SellButton), {
  ssr: false,
  loading: () => <></>,
});

```

Leave it to stay only on `Sell Btn` for comparison and Sell btn with this config is alright but `Buy` with exact same state as Sell, just `ssr: true` and stays disabled :X

**Reply by icyJoseph** [reply]
Using next/dynamic should not even be considered here. Let's stay focused on the root issue.

How is your hook working internally?

This is really odd. Got a link to a repository/site? 

What Node.js version do you have? Do you see this in production too? Only dev?

**Reply by icyJoseph** [reply]
I've tried to reproduce this issue, but it works fine... :/ are there any more logs or something look at?

**Reply by AndonMitev** [reply]
node --v: `v18.17.1`
env: dev && prod
"next": "14.0.3",

Repo currently is not open source :X but this is the hook:

```
import { useAccount, useNetwork } from 'wagmi';

export function useIsChainSupported() {
  const { connector } = useAccount();
  const { chain } = useNetwork();

  const supportedChainsId = connector?.chains?.map((chain) => chain.id) ?? [];
  return supportedChainsId.includes(chain?.id || 0);
}

```

I'll try to provide related components, structure goes like this:

`/app/trends/page.tsx` - https://pastebin.com/4A4aY9SW, from here we enter in line 41, `Trends table`:
`trendsTable`: - https://pastebin.com/tXKryWy9, this is where `Buy` || `Sell` btn shows up:
`Buy Btn`: - https://pastebin.com/6t4JGrT4, this is where Button shows up:
`Button`: - https://pastebin.com/KKvdLCui, thats the whole flow idk if that helps somehow 

Something additional that i have spot on, is not disabled everytime, sometimes they are enabled and sometimes they are disabled but is pretty easy to reproduce like 50% of the times will be disabled when actually has to be enabled
